includes/class-freemius.php

```diff
@@ -19058,6 +19316,204 @@ private function get_complete_upgrade_instructions( $plan_title = '' ) {
+                    "<small class='products'><strong>Products:</strong> %s</small>",
+                    $products_and_add_ons
+                );
+            }
+
+            $actions = sprintf(
+                '<ul><li>%s<span class="action-description"> - send me security & feature updates, educational content and offers.</span></li><li>%s<span class="action-description"> - do <span class="underline">NOT</span> send me security & feature updates, educational content and offers.</span></li></ul>',
+                sprintf('<button class="button button-primary allow-marketing">%s</button>', $this->get_text_inline( 'Yes' ) ),
+                sprintf('<button class="button button-secondary">%s</button>', $this->get_text_inline( 'No' ) )
+            );
+
+            return sprintf(
+                '%s %s %s',
+                $thank_you,
+                $already_opted_in,
+                $this->get_text_inline( 'Due to the new GDPR compliance requirements it is required that you provide your explicit consent, again, confirming that you are onboard 🙂') .
+                '<br />' .
+                $this->get_text_inline( "Please let us know if you'd like us to contact you for security & feature updates, educational content, and occasional offers:" ) .
+                $actions .
+                ( $is_single_parent_product ? '' : $multiple_products_text )
+            );
+        }
+
+        /**
+         * @author Leo Fajardo (@leorw)
+         * @since 2.1.0
+         *
+         * @param string $user_email
+         * @param array  $plugin_ids
+         *
+         * @return array
+         */
+        private function get_user_plugins( $user_email, $plugin_ids ) {
+            $url = WP_FS__ADDRESS . '/action/service/user_plugin/';
+
+            if ( WP_FS__DEBUG_SDK || isset( $_COOKIE['XDEBUG_SESSION'] ) ) {
+                $url = add_query_arg( 'XDEBUG_SESSION_START', rand( 0, 9999999 ), $url );
+                $url = add_query_arg( 'XDEBUG_SESSION', 'PHPSTORM', $url );
+
+                $request['cookies'] = array(
+                    new WP_Http_Cookie( array(
+                        'name'  => 'XDEBUG_SESSION',
+                        'value' => 'PHPSTORM',
+                    ) )
+                );
+            }
+
+            $params = array(
+                'email'      => $user_email,
+                'plugin_ids' => $plugin_ids
+            );
+
+            $request = array(
+                'method'  => 'POST',
+                'body'    => $params,
+                'timeout' => WP_FS__DEBUG_SDK ? 60 : 30,
+            );
+
+            $result = array();
+
+            $total_plugin_ids             = count( $plugin_ids );
+            $plugin_ids_count_per_request = 10;
+            for ( $i = 1; $i <= $total_plugin_ids; $i += $plugin_ids_count_per_request ) {
+                $plugin_ids_set                = array_slice( $plugin_ids, $i - 1, $plugin_ids_count_per_request );

```

Comment:
@fajardoleo since this method isn't using our API layer, we need to add some caching to avoid a situation where for some reason this request is called multiple times in a short period. I think that transients are the easiest here. So you can do something like that:
```php
    $cache_key = 'fs_user_plugins_' . md5( $user_email, implode( ',', $plugin_ids_set) );
    $response = get_transient( $cache_key );

    if ( false === $response ) {
        ... actual API request ...
        set_transient( $cache_key, $response, WP_FS__TIME_24_HOURS_IN_SEC);
    }

    ... a logic that processes the response...
```